

### 异常

#### 1. 什么是 Java 异常？

异常是发生在程序执行过程中阻碍程序正常执行的错误事件。比如：用户输入错误数据、硬件故障、网络阻塞都会导致出现异常。只要在 Java 语句执行中产生了异常，一个异常对象就会被创建，JRE 就会试图寻找异常处理程序来处理异常。如果有合适的异常处理程序，异常对象就会被异常处理程序接管，否则，将引发运行环境异常，JRE 终止程序执行。Java 异常处理框架智能处理运行时错误，编译错误不在其考虑范围之内。

#### 2. Java 异常处理中有哪些关键字？

- throw：有时我们需要显式地创建并抛出异常对象来终止程序的正常执行。throw 关键字用来抛出并处理运行时异常。
- throws：当我们抛出任何「被检查的异常(checked exception)」并不处理时，需要在方法签名中使用关键字 throws 来告知调用程序此方法可能会抛出的异常。调用方法可能会处理这些异常，或者同样用 throws 来将异常传给上一级调用方法。throws 关键字后可接多个潜在异常，甚至是在 main() 中也可以使用 throws。
- try-catch：在代码中用 try-catch 块处理异常。当然，一个 try 块之后可以有多个 catch 子句，try-catch 块也能嵌套。每个 catch 块必须接受一个（且仅有一个）代表异常类型的参数。
- finally：finally 块是可选的，并且只能配合 try-catch 一起使用。虽然异常终止了程序的执行，但是还有一些打开的资源没有被关闭，因此，我们能使用 finally 进行关闭。不管异常有没有出现，finally 块总会被执行。

#### 3. 描述一下异常的层级

Java 异常是层级的，并通过继承来区分不同种类的异常。

- Throwable 是所有异常的父类，它有两个直接子对象 Error 与 Exception，其中 Exception 又被继续划分为「被检查的异常(checked xception)」和「运行时的异常(Runtime Exception,即不受检查的异常)」。 Error 表示编译时和系统错误，通常不能预期和恢复，比如硬件故障、JVM 崩溃、内存不足等。
- 被检查的异常(checked exception)在程序中能预期，并要尝试修复，如 FileNotFoundException。我们必须捕获此类异常，并为用户提供有用信息和合适日志来进行调试。Exception 是所有被检查的异常的父类。
- 运行时异常(Runtime Exception)又称为不受检查异常，源于糟糕的编程。比如我们检索数组元素之前必须确认数组的长度，否则就可能会抛出 ArrayIndexOutOfBoundException 运行时异常。RuntimeException 是所有运行时异常的父类。

#### 4. 什么是异常链？

在进行一个异常处理时抛出了另外一个异常，由此产生了一个异常链条。如果因为因为异常你决定抛出一个新的异常，一定要包含原有的异常，这样，处理程序才可以通过 getCause() 和 initCause() 方法来访问异常最终的根源。

#### 5. final、finalize 与 finally 的区别？

final 和 finalize 是 Java 的关键字，而 finalize 是方法。
- final：在创建不可变的类的时候非常有用，只是声明这个类是 final 的。
- finally：异常处理的关键字，关闭连接和资源文件的时候需要用到 finally 块。
- finalize()：垃圾回收器在回收一个对象前调用。

#### 6. 异常处理的时候，finally 代码块的重要性是什么？

无论是否抛出异常，finally 代码块总是会被执行（除非在 finally 块之前执行了 `System.exit()`，退出虚拟机）。finally 代码块主要用来释放资源，比如：I/O 缓冲区，数据库连接等。